src/v3/plugins/github/graphView.js

```diff
@@ -0,0 +1,298 @@
+    return this._children(repo, GN._Prefix.issue);
+  }
+
+  pulls(repo: GN.RepoAddress): Iterator<GN.PullAddress> {
+    return this._children(repo, GN._Prefix.pull);
+  }
+
+  comments(commentable: GN.CommentableAddress): Iterator<GN.CommentAddress> {
+    return this._children(commentable, GN._Prefix.comment);
+  }
+
+  reviews(pull: GN.PullAddress): Iterator<GN.ReviewAddress> {
+    return this._children(pull, GN._Prefix.review);
+  }
+
+  // TODO(@wchrgin) figure out how to overload this fn signature
+  parent(child: GN.ChildAddress): GN.ParentAddress {
+    const options = {
+      direction: Direction.OUT,
+      edgePrefix: GE._Prefix.hasParent,
+      nodePrefix: GN._Prefix.base,
+    };
+    const parents: GN.ParentAddress[] = Array.from(
+      this._neighbors(child, options)
+    );
+    if (parents.length !== 1) {
+      throw new Error(
+        `Parent invariant violated for child: ${stringify(child)}`
+      );
+    }
+    return parents[0];
+  }
+
+  authors(content: GN.AuthorableAddress): Iterator<GN.UserlikeAddress> {
+    const options = {
+      direction: Direction.IN,
+      edgePrefix: GE._Prefix.authors,
+      nodePrefix: GN._Prefix.userlike,
+    };
+    return this._neighbors(content, options);
+  }
+
+  _maybeCheckInvariants() {
+    if (this._isCheckingInvariants) {
+      return;
+    }
+    if (process.env.NODE_ENV === "test") {
+      // TODO(perf): If this method becomes really slow, we can disable
+      // it on specific tests wherein we construct large graphs.
+      this.checkInvariants();
+    }
+  }
+
+  checkInvariants() {
+    this._isCheckingInvariants = true;
+    try {
+      this._checkInvariants();
+    } finally {
+      this._isCheckingInvariants = false;
+    }
+  }
+
+  _checkInvariants() {
+    // Check the node type <-> prefix mapping is maintained
+    const nodeTypeToInvariant = {
+      [GN.REPO_TYPE]: {hasParent: false, parentAccessor: null},
+      [GN.ISSUE_TYPE]: {hasParent: true, parentAccessor: (x) => x.repo},
+      [GN.PULL_TYPE]: {hasParent: true, parentAccessor: (x) => x.repo},
+      [GN.COMMENT_TYPE]: {hasParent: true, parentAccessor: (x) => x.parent},
+      [GN.REVIEW_TYPE]: {hasParent: true, parentAccessor: (x) => x.pull},
+      [GN.USERLIKE_TYPE]: {hasParent: false, parentAccessor: null},
+    };
+    for (const node of this._graph.nodes({prefix: GN._Prefix.base})) {
+      const structuredNode = GN.fromRaw((node: any));
+      const type = structuredNode.type;
+      const invariant = nodeTypeToInvariant[type];
+      if (invariant == null) {
+        throw new Error(`Invariant: Unexpected node type ${type}`);
+      }
+      const {hasParent, parentAccessor} = invariant;
+      if (hasParent) {
+        // Will throw invariant error if there is not exactly 1 parent
+        const parent = this.parent((structuredNode: any));
+        if (parentAccessor == null) {
+          throw new Error(`Impossible; make flow happy`);
```

Comment:
Minor complaint (but not, I think, a nit): this is impossible in a way
that could be changed in the future such that this error is thrown at
runtime\*. I suggest either of the following alternatives:

  - Make the message clearer about what has gone wrong: this is an error
    in the invariant-checking code itself. For instance, something like
    `bad invariant for ${type}: missing accessor` would make me happier.

  - Better yet, you can encode this check statically. The problem is
    that your invariants are encoded as

    ```js
    type Invariant = {|
      +hasParent: boolean,
      +parentAccessor: ?((*) => GN.ParentAddress),
    |};
    ```

    where there is a (meta-)invariant between the value of the first
    field and the nullity of the second that is not enforced by the
    type. Instead, you could use either a union of more specific object
    types:

    ```js
    type Invariant =
      | {|+hasParent: true, +parentAccessor: (*) => GN.ParentAddress|}
      | {|+hasParent: false, +parentAccessor: null|};
    ```

    or simply functions:

    ```js
    type Invariant = ?((*) => GN.ParentAddress);
    ```

    I’ve verified that each of these typechecks without the
    `null`-check, and also that the first option statically errors if
    the meta-invariant is violated (in the second option, it’s not
    possible to write something that violates this meta-invariant).

\* What I mean is that the impossibility is [admissible but not
derivable][1], which I think is an important didstinction.

[1]: https://en.wikipedia.org/wiki/Rule_of_inference#Admissibility_and_derivability